cpp/include/cugraph/detail/shuffle_wrappers.hpp

```diff
@@ -277,4 +270,45 @@ rmm::device_uvector<value_t> collect_local_vertex_values_from_ext_vertex_value_p
   bool do_expensive_check);
 
 }  // namespace detail
+
+/**
+ * @ingroup shuffle_wrappers_cpp
+ * @brief Shuffle keys and their associated properties to the proper GPU based on a partitioning
+ * function.  Keys for this function are assumed to be an arithmetic type
+ *
+ * @tparam key_t type of key
+ * @tparam key_to_gpu_op_t Function to convert key to GPU id
+ *
+ * @param[in] handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator,
+ * and handles to various CUDA libraries) to run graph algorithms.
+ * @param [in] key Device vector of keys
+ * @param [in] properties Vector of device vectors of properties associated with each key
+ * @param [in] key_to_gpu_op function converting key to a GPU id
+ *
+ * @return tuple of device vector of keys and vector of device vector of properties associated with
+ * the key.
+ */
+template <typename key_t, typename key_to_gpu_op_t>
+std::tuple<rmm::device_uvector<key_t>, std::vector<arithmetic_device_uvector_t>>
+shuffle_keys_with_properties(raft::handle_t const& handle,
+                             rmm::device_uvector<key_t>&& keys,
+                             std::vector<arithmetic_device_uvector_t>&& properties,
+                             key_to_gpu_op_t key_to_gpu_op);
```

Comment:
We have cugraph/shuffle_functions.hpp and cugraph/detail/shuffle_wrappers.hpp, should we keep both?

I am thinking about moving non-detail space functions to shuffle_functions.hpp and create a non-detail space functions for functions in the detail/shuffle_wrappers.hpp and eventually removing shuffle_wrappers.hpp.

What do you think?

Reply:
I was thinking along similar lines.  I was also thinking of moving the detail functions out of the include directory and into the src directory to remove them more completely from the public API.  I have more shuffle changes coming in the next batch of changes... I'll try moving some files around then.